### DISCUSSION TITLE
Timing Request Durations Correctly

### DISCUSSION THREAD

**Discussion by darraghjones** [dotnet/aspnetcore#23664] [General]
Given the following most basic of ASP.NET Core applications:

```
    public class Program
    {
        public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.Configure(appBuilder =>
                        appBuilder.Run(async context =>
                        {
                            var stopwatch = Stopwatch.StartNew();
                            Thread.Sleep(1000);
                            await context.Response.WriteAsync($"Finished in {stopwatch.ElapsedMilliseconds} milliseconds.");
                        }));
                });
    }
```

And the following appsettings.json

```
{
  "Logging": {
    "LogLevel": {
      "Default": "None",
      "Microsoft.AspNetCore.Hosting.Diagnostics" : "Information",
    }
  },
  "AllowedHosts": "*"
}
```

If I run an even moderate load test (100 requests, using bombardier in my case) I see latency of around 5 seconds.

```
 ~/go/bin/bombardier http://localhost:5000 -l -n 100 -t 60s
Bombarding http://localhost:51568 with 100 request(s) using 125 connection(s)
 100 / 100 [=================================================================================================================================] 100.00% 16/s 6s
Done!
Statistics        Avg      Stdev        Max
  Reqs/sec        19.46     250.28    4086.58
  Latency         5.21s   366.21ms      6.05s
  Latency Distribution
     50%      5.05s
     75%      5.05s
     90%      6.04s
     95%      6.05s
     99%      6.05s
  HTTP codes:
    1xx - 0, 2xx - 100, 3xx - 0, 4xx - 0, 5xx - 0
    others - 0
  Throughput:     3.31KB/s
```

However, all I see in the logs are

> info: Microsoft.AspNetCore.Hosting.Diagnostics[2]
      Request finished in 1003.3658ms 200 

Clearly the requests are taking more than 1 second. I believe the unaccounted 4 seconds are when the request is queued on the ThreadPool. 

So my question is how can I measure this latency from _inside_ my application?

**Comment by Tratcher** [comment]
You're starving the thread pool. Change the thread sleep to an async task delay. You don't get more accurate data because the requests are stuck down at the network layer.

**Reply by darraghjones** [reply]
Thanks for your response. I know that if I change the Thread.Sleep to an await Task.Delay this will improve the performance. But in a real application, it might not be as simple as this. 

Can you elaborate on what you mean by the "_requests are stuck down at the network layer"_?

If I run `dotnet counters monitor --process-id <processId>` I see the output below. Notice in particular that the **ThreadPool Queue Length** is 65. I believe this is where the requests are *stuck* and are therefore available to the dotnet runtime to starting timing the request duration.

```
Press p to pause, r to resume, q to quit.
    Status: Running

[System.Runtime]
    % Time in GC since last GC (%)                         0
    Allocation Rate / 1 sec (B)                      168,272
    CPU Usage (%)                                          0
    Exception Count / 1 sec                                0
    GC Heap Size (MB)                                      6
    Gen 0 GC Count / 60 sec                                0
    Gen 0 Size (B)                                         0
    Gen 1 GC Count / 60 sec                                0
    Gen 1 Size (B)                                         0
    Gen 2 GC Count / 60 sec                                0
    Gen 2 Size (B)                                         0
    LOH Size (B)                                           0
    Monitor Lock Contention Count / 1 sec                  1
    Number of Active Timers                                1
    Number of Assemblies Loaded                           87
    ThreadPool Completed Work Item Count / 1 sec          18
    ThreadPool Queue Length                               65
    ThreadPool Thread Count                               18
    Working Set (MB)                                      51
```

**Reply by Tratcher** [reply]
You'd have to look in the debugger to find out what those queued tasks are. They are likely the server tasks waiting to read the next request. The request timer doesn't start until the server has fully received and dispatched the request to the application.

**Reply by darraghjones** [reply]
Presumably some event triggers the server to create a task to read the next request? I'd like to start the request duration at this point....before it has been dispatched to the application.

**Reply by Tratcher** [reply]
That gets deep into the server implementation details. Which server are you using here? Kestrel, HttpSys, IIS in-proc, or IIS out-of-proc?

**Reply by darraghjones** [reply]
In my case I'm using Kestrel.

**Reply by Tratcher** [reply]
@davidfowl @shirhatti these are metrics you've been considering in.

**Reply by davidfowl** [reply]
These metrics will need to be added in the various server implementations. I have an idea on how we can do this cheaply. We should expose a timestamp on the HttpContext that the server implementation can set that will determine the true start time of the request.

@darraghjones There are 2 things you can measure:
- Time until the response is generated
- Time until the code running on the server is over.

Which do you care about?

**Reply by darraghjones** [reply]
I guess what I care about is the worst possible experience for the end-user. So, out of those two cases, I expect the latter would be the closest to this.

**Reply by darraghjones** [reply]
Any chance these metrics will make it into dotnet 5?

**Reply by davidfowl** [reply]
.NET 5 is done so no.

**Reply by darraghjones** [reply]
i've retested the scenario, this time using the opentelemetry metrics. i can see the metric: http.server.request.duration suffers the same issue. 

![image](https://github.com/user-attachments/assets/23c13a28-265a-43ba-8aad-ac6709c279f8)
